nixos/modules/services/abstract/generic.nix

```diff
@@ -0,0 +1,23 @@
+# This module is valid for all implementations of the service abstraction layer,
+# and might even be vendored from another repo or moved outside of nixpkgs/nixos/
+# into nixpkgs/something.
+
+{ lib, config, options, pkgs, ... }:
+let
+  inherit (lib) mkOption types;
+in
+{
+  options = {
+    services.abstract = mkOption {
+      type = types.lazyAttrsOf (types.submoduleWith {
+        modules = [
+          ./instance/generic.nix
```

Comment:
This hardcodes the generic options, but that's not actually necessary and creates some confusion about what the actually used interface really is, in a way that isn't forward compatible.

It should be up to the actual service to decide in which format it wants to be loaded, and process managers can decide which formats they support.

This could be negotiated through specialArgs and may look like:

```nix
# postgresql.nix
{ serviceInterface, ... }: {
  imports = [
    serviceInterface.generic_1_0
    serviceInterface.systemd_nixos_23_11 or { }
  ];
  options = {
    # ...
  };
}
```

This says that a postgresql service expects the process manager to support the `generic_1_0` interface and load that.
The `systemd_1_0` interface is optional and is only loaded if the process manager supports it.

`imports` can of course only depend on `specialArgs` arguments, but that is sufficient for this purpose.

Note that services can already introspect e.g. `options?systemd` to figure out in which kind of process manager they are loaded. So the addition of such a mechanism is primarily important for forward compatibility as we'll inevitably evolve the taxonomy of services.

This also allows the creation of helper modules that for instance provide polyfills, emulate one process manager's features on top of another, etc. These could be distributed independently, or as part of the suite of modules and interfaces that are built in to the process manager integrations.

Even just within the NixOS context, such versions let use eat the cake and have it too:
 - NixOS/systemd can make breaking changes in new versions
 - third party service modules (e.g. flakes, etc) have stable targets that they can support

Arguably there's a non-zero cost to having a compatibility layer or keeping old code around, but at least it becomes a solvable problem, and the old solution doesn't come with the expectation of parity with the new solution.